gobblin-runtime/src/main/java/org/apache/gobblin/runtime/GobblinMultiTaskAttempt.java

```diff
@@ -416,14 +425,17 @@ private boolean taskSuccessfulInPriorAttempt(String taskId) {
           // task could not be created, so directly count down
           countDownLatch.countDown();
           log.error("Could not create task for workunit {}", workUnit, e);
-        } else if (!task.hasTaskFuture()) {
+        } else if (taskFuture == null) {
           // Task was created and may have been registered, but not submitted, so call the
           // task state tracker task run completion directly since the task cancel does nothing if not submitted
           this.taskStateTracker.onTaskRunCompletion(task);
           log.error("Could not submit task for workunit {}", workUnit, e);
         } else {
-          // task was created and submitted, but failed later, so cancel the task to decrement the CountDownLatch
-          task.cancel();
+          // task was created and submitted, but failed later(and still pop back here), so cancel the task to decrement the CountDownLatch
+          if (!taskFuture.cancel(true)) {
+            log.error("Failure in terminating submitted task");
+          }
+          this.taskStateTracker.onTaskRunCompletion(task);
```

Comment:
Do we want to call `this.taskStateTracker.onTaskRunCompletion(task)` if `taskFuture.cancel(true)` returns false?